### PR TITLE
`Development`: Refactor isAtLeast occurrences in client and unify spy testing

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.

### Motivation and Context

#3742 refactored the client code such that exercise/course objects store the current access rights in the client. There were not used in some components.

Additionally GitHub was complaining about the checking of some spies in the client. These are fixed and unified in the code as well. (Point 11: https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/)

### Description

This PR refactores the last components to also use the exercise/course objects for the access rights.
The occurrences using the Metis service (e.g. post-header.component) is intentionally left out.

For checking the spies, we use .toHaveBeenCalledOnce()/.toHaveBeenCalledTimes(...) instead of .toHaveBeenCalled() and .not.toHaveBeenCalled() instead of .toHaveBeenCalled(0). This is unified in this PR

#### Code Review
- [x] Review 1
- [x] Review 2

### CODE REVIEW

File: src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard-information.component.html
Comment (Strohgelaender) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard-information.component.html:47]: This would evaluate to true of the course is not defined! Please check if this could cause any problems.

Comment (JohannesStoehr) [reply src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard-information.component.html]: Good catch. I also improved such occurrences in three other files

File: src/main/webapp/app/assessment/assessment-header/assessment-header.component.html
Comment (Strohgelaender) [inline src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:35]: Now the questionmark is no longer needed here.
```suggestion
                !exercise.isAtLeastInstructor &&
```

**Review by Strohgelaender — APPROVED** [review]
code looks good

**Review by Dominik-Weinzierl — DISMISSED** [review]
Looks good to me :+1:

**Review by minrows — APPROVED** [review]
Refactored code LGTM 👍

**Review by babyygemperor — APPROVED** [review]
Code looks good, reviewed in a testing session 👍🏼

**Review by ge65cer — APPROVED** [review]
Approve code changes.

**Comment by ge65cer** [comment]
I requested an Eslint rule for your changes over at https://github.com/jest-community/eslint-plugin-jest/issues/1198.

File: src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html
Comment (julian-christl) [inline src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html:100]:
```suggestion
        [readOnly]="readOnly || (exercise && !exercise.isAtLeastEditor && !toComplete)"
```
Wouldn't that be enough?

**Review by julian-christl — CHANGES_REQUESTED** [review]
Two things I would like to have you look at.

And another thing:
What about the other occurrences of `toHaveBeenCalled()`? Do you plan to do them in another PR? Cause otherwise, I'm not quite sure why you would only replace them partially.
There are also four occurrences remaining of `toHaveBeenCalledTimes(0)`.

File: src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html:100]: This can be a problem when exercise has the type Exercise | undefined.
The double !! makes sure that the types are correct, since this will always produce a boolean

Comment (julian-christl) [reply src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html:100]: That's right. Maybe I understand wrongly how that resolves. My understanding is:

exercise of type Exercise => is an object => `exercise` (true-ish); `!exercise` (false); `!!exercise` (true)
exercise undefined => `exercise` (false-ish); `!exercise` (true); `!!exercise` (false)

In my opinion, this should be enough. Where do you see the problem? Or does the linter complain? That's another story.

Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.html:100]: Exactly, the linter does not like it without the !!

**Review by ge65cer — DISMISSED** [review]
Re-approve after merge.

**Review by julian-christl — APPROVED** [review]
Changes consistent, no occurrences in the IDE, code approved.

**Comment by krusche** [comment]
We need more approvals, one is not enough!

**Review by ge65cer — APPROVED** [review]
Tested locally, found no issues.

**Review by Strohgelaender — APPROVED** [review]
code looks good. I did not find any other vialations in the code.

**Review by julian-christl — APPROVED** [review]
Reapprove code. New changes consistent

**Review by manuelmanso — APPROVED** [review]
Code looks good to me

**Review by ge65cer — APPROVED** [review]
Re-approve after merge

**Pull request merged**
